Initial Validator (commit case)

KtorZ opened the pull request.

ch1bo committed:
Add a healthyCommitTx and validate it

Factored out 'mkInitialOutput' to be able to use it in the commit unit tests.

ch1bo committed:
Improve property output by adding BasicFailure as counter example

ch1bo committed:
Extend describeCardanoTx by printing inputs

ch1bo committed:
Add committed utxo to the lookup utxo

ch1bo committed:
Small refactor on mkInitialOutput

ch1bo committed:
Fix healthyCommitTx to use the same PubKeyHash

ch1bo committed:
Add Commit mutator for changing committed value

ch1bo committed:
Made property output better understandable again

ch1bo committed:
Add output values to describeCardanoTx

ch1bo committed:
Have Initial validator check values

ch1bo committed:
Make committed value const 8₳ in healthyCommitTx

ch1bo committed:
Implement a debugInteger to show decimals

ch1bo committed:
Fix/fake initial validator to check comittedValue

Fake it till you make it.. how would we catch this hard-coded
committedValue with only a single healthyCommitTx?

ch1bo committed:
Fix imports

ch1bo committed:
Keep abortTx valid by distinguishing initial redeemer as Commit | Abort

ch1bo committed:
Add a Maybe TxOutRef to Initial.Commit redeemer

Also adds another isomorphism for Plutus.TxOutRef

ch1bo committed:
Lookup actual comittedValue in Initial validator

ch1bo committed:
Draft MutateComittedValue mutation

KtorZ committed:
Implement mutation changing commit value during initial script validation.

Note that, we could also change the datum (or more generally, the resolving input associated with that commit).

ch1bo committed:
Add MutateCommittedAddress mutator

Also adds some helpers to Hydra.Ledger.Cardano

abailly committed:
Check Initial script datum in the case of no committed UTXO

abailly committed:
Check committed TxOut is preserved in commit datum

KtorZ committed:
Start removing SerializedTxOutRef from commit validator's datum

  Not needed, but was handy for observing commit. Removing it makes on-chain validations easier / less error prone.

ch1bo committed:
Implement observeCommitTx according to new strategy

abailly committed:
Make test compile passing the list of inputs

KtorZ committed:
Correctly deserialize initial redeemer when observing commit.

ch1bo commented:
Will have a go on refactoring our changes from the ensemble without changing the semantics to make this more palatable.

ch1bo committed:
Refactor checkComittedDatum

When trying to factor out the datum lookup, I realized that the commit
validator datum might not be part of the commit transaction, i.e. it's
optional as far as the ledger is concerned.

ch1bo committed:
Resolve/update TODOs

ch1bo committed:
Small refactor in observeCommitTx

ch1bo committed:
Re-introduce Initial.compiledValidator

Likely got lost during rebase
